### ISSUE TITLE
Privacy: Session Storage and Persistence section should apply to all CDM storage

### ISSUE THREAD

**Issue by ddorwin** [w3c/encrypted-media#169]
While working on #117, I found that most of our requirements for CDM storage are in the [Session Storage and Persistence](https://w3c.github.io/encrypted-media/#session-storage) section, which is a subsection of the `MediaKeySession` Interface.

The same restrictions (per-origin, etc.) should apply to all CDM storage, not just those associated with a `MediaKeySession` object. Specifically, if a CDM instance needs to store data unrelated to a specific session or related to multiple sessions, the same requirements should apply. As with session-specific data, such data must be fully clearable. (See also #168.)

We should also be explicit that implementations MUST NOT store data outside the context of these object, especially in a non-origin-bound way.

**Comment by steelejoe** [comment]
Just to be clear, you are saying that all data stored by the CDM should be origin-specific and clearable, You are not saying that all data stored by the CDM has to be associated with a specific session. Is that correct?

**Comment by ddorwin** [comment]
@steelejoe: Yes.

**Comment by steelejoe** [comment]
Great. Then this sounds like a good idea. I can put together a pull request for us to discuss. I am thinking a new storage section under 5 MediaKeys, and then reference back to that from the current MediaKeySession section.

**Comment by mwatson2** [comment]
@steelejoe Any progress on a PR for this one ?

**Comment by jdsmith3000** [comment]
@ddorwin Is the scope of this topic broader than session data (already specified) and Distinctive Identifiers (presumably to be covered under #117)?  Are there known cases where a `MediaKey` may store data outside that covered by the `sessionTypes` under `MediaKeySession`?

**Comment by ddorwin** [comment]
@jdsmith3000, yes, the purpose of this bug is to apply the same requirements we apply to persistent session storage to any and all data stored by the CDM. As I said in https://github.com/w3c/encrypted-media/issues/57#issuecomment-213643420:
> Persistent sessions are one type of persistent state, but we do not limit `persistentState` to only sessions. The implementation expectations for storage, clearing, etc. of all persistent state is the same, though. (See #169.)

There are no normative cases where `MediaKeys` algorithms store data, but CDM instances are permitted to store per-origin clearable data if the `MediaKeys` object was created by a `MediaKeySystemAccess` object with `persistentState` = `"required"`.

**Comment by ddorwin** [comment]
On a related note, `createMediaKeys()` algorithm step 2.9 says:
>If persistent state allowed is false, prevent instance from persisting any state related to the application or origin of this object's Document.

This implies that **non-origin-specific** state may be persisted even if `persistentState` = `"not-allowed"`.

I think we might assume in various places that this is not allowed, and it might be laborious to address it everywhere we need to. If there are no known cases for this, maybe we should reword this to say "prevent instance from persisting any state" then follow that with a statement that state must be per-origin.

The one case I can think of where there might be non-origin-specific state is some intermediary step in individualization. For example, some one-time processing of a secret before acquiring the per-origin Distinctive Identifier. We could explicitly address that in the text, but then there are questions about clearability. Perhaps this is part of the platform and not the CDM instance.

**Comment by mwatson2** [comment]
@ddorwin Non-origin-specific state at some intermediary step in individualization would not be _persistent_ state, would it ?

Another question: if `MediaKeys` is created with `persistentState` = `not-allowed` is the CDM allowed access to _read_ persistent state (origin-specific or not) that was created by a previous `MediaKeys` instance created with `persistentState` = `required` ? I would imagine not.

**Comment by mwatson2** [comment]
PR for discussion: #191.

**Comment by mwatson2** [comment]
Reviewed at editors' meeting (@mwatson2 @jdsmith3000). No comments.

**Comment by ddorwin** [comment]
>@ddorwin Non-origin-specific state at some intermediary step in individualization would not be _persistent_ state, would it ?

It depends on whether it is stored in memory or something non-ephemeral. (In this case, I was thinking that the pre-processing is persisted to later be used for each origin.) It is not `persistentState`.

>Another question: if `MediaKeys` is created with `persistentState` = `not-allowed` is the CDM allowed access to _read_ persistent state (origin-specific or not) that was created by a previous `MediaKeys` instance created with `persistentState` = `required` ? I would imagine not.

I agree that it should not. The exception would be things related to `distinctiveIdentifier` and thus explicitly excluded from `persistentState`. This includes both a persisted identifier resulting from an individualization process and, for example, the intermediary state discussed above. All of this must be clearable as defined by the relevant sections on Distinctive Identifiers.


<br/>Applying that to the intermediary state, it seems this one known possible exception should be handled as Distinctive Identifier (hopefully #117 will address this) and we can require that all "persistent state" covered by `persistentState` be per-origin.

**Comment by mwatson2** [comment]
@ddorwin @mwatson2 @jdsmith3000 reviewed. This is ready to merge once the typo noted in the PR line notes is addressed.
